packages/ckeditor5-widget/src/widgettypearound/widgettypearound.js

```diff
@@ -0,0 +1,250 @@
+				injectUIIntoWidget( editor.editing.view, buttonTitles, viewElement );
+			}
+		}, { priority: 'low' } );
+	}
+
+	/**
+	 * Registers an editing view post-fixer which checks all block widgets in the content
+	 * and adds CSS classes to these which should have the typing around (UI) enabled
+	 * and visible for the users.
+	 *
+	 * @private
+	 */
+	_enableDetectionOfTypeAroundWidgets() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const editingView = editor.editing.view;
+
+		function positionToWidgetCssClass( position ) {
+			return `ck-widget_can-type-around_${ position }`;
+		}
+
+		editingView.document.registerPostFixer( writer => {
+			// Find all view elements in the editing root.
+			[ ...editingView.createRangeIn( editingView.document.getRoot() ) ]
+				// ...then filter only the widgets that may need the type around feature.
+				.filter( ( { item: widgetViewElement } ) => {
+					const modelElement = editor.editing.mapper.toModelElement( widgetViewElement );
+
+					return isTypeAroundWidget( widgetViewElement, modelElement, schema );
+				} )
+				// ...and update widgets' classes depending on possible positions for paragraph insertion.
+				.forEach( ( { item: widgetViewElement } ) => {
+					const positions = getWidgetTypeAroundPositions( widgetViewElement );
+
+					// Remove all classes. In theory we could remove only these that will not be added a few lines later,
+					// but since there are only two... KISS.
+					writer.removeClass( POSSIBLE_INSERTION_POSITIONS.map( positionToWidgetCssClass ), widgetViewElement );
+
+					// Set CSS classes related to possible positions. They are used so the UI knows which buttons to display.
+					writer.addClass( positions.map( positionToWidgetCssClass ), widgetViewElement );
+				} );
+		} );
+	}
+
+	/**
+	 * Registers a `mousedown` listener for the view document which intercepts events
+	 * coming from the type around UI, which happens when a user clicks one of the buttons
+	 * that insert a paragraph next to a widget.
+	 *
+	 * @private
+	 */
+	_enableInsertingParagraphsOnButtonClick() {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+
+		editingView.document.on( 'mousedown', ( evt, domEventData ) => {
+			const button = getClosestTypeAroundDomButton( domEventData.domTarget );
+
+			if ( !button ) {
+				return;
+			}
+
+			const buttonPosition = getTypeAroundButtonPosition( button );
+			const widgetViewElement = getClosestWidgetViewElement( button, editingView.domConverter );
+
+			this._insertParagraph( widgetViewElement, buttonPosition );
+
+			domEventData.preventDefault();
+			evt.stop();
+		} );
+	}
+}
+
+// Injects the type around UI into a view widget instance.
+//
+// @param {module:engine/view/view} editingView
+// @param {Object.<String,String>} buttonTitles
+// @param {module:engine/view/element~Element} widgetViewElement
+function injectUIIntoWidget( editingView, buttonTitles, widgetViewElement ) {
```

Comment:
You could replace editing view here with the writer that you can take from conversionApi (in the converter event).